### PR TITLE
BREAKING CHANGE: support koreanbots v2

### LINKED ISSUE #4: Supporting v2 API
# Should support v2 API(GraphQL) before v2 release.

Working at #6

### BaseURL
Beta(For now):

```
beta.koreanbots.dev/api
```

Production
```
koreanbots.dev/api
```

### Document

https://github.com/koreanbots/api-docs

## To-do

- [ ] Get User
- [ ] Get Bot
- [ ] Update Server

### PR DESCRIPTION
**Pull request by SaidBySolo**
closes #4 

Ref: https://koreanbots.dev/developers/docs

### CODE REVIEW

File: koreanbots/typing.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    "도박",
+    "게임",
+    "밈",
+    "레벨링",
+    "유틸리티",
+    "번역",
+    "대화",
+    "NSFW",
+    "검색",
+]
```
Comment (wonderlandpark) [inline koreanbots/typing.py:18]: Categories are added please check docs

Ref: https://beta.koreanbots.dev/developers/docs/엔드포인트/봇#category

File: koreanbots/http.py [hunk 2]
```diff
@@ -1,355 +1,65 @@
...
+                        response.status, await response.json()
+                    )
+                else:
+                    raise HTTPException(response.status, await response.json())
+
+            return await getattr(response, return_method)()
+
+    @required
+    async def get_vote(self, user_id: int):
+        await self.request("GET", "json", f"/bots/voted/{user_id}")
```
Comment (wonderlandpark) [inline koreanbots/http.py:47]: This would change, just leaving a comment for reminding

**Review by wonderlandpark — CHANGES_REQUESTED** [review]

File: koreanbots/client.py [hunk 3]
```diff
@@ -1,251 +1,49 @@
...
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
-"""
+import asyncio
+from koreanbots.decorator import strict_literal
```
Comment (kijk2869) [inline koreanbots/client.py:2]: Make it dynamic import

File: koreanbots/client.py [hunk 4]
```diff
@@ -1,251 +1,49 @@
...
+from typing import Literal, Optional, Type
 
+import aiohttp
+from koreanbots.http import KoreanbotsRequester
+from koreanbots.errors import HTTPException
```
Comment (kijk2869) [inline koreanbots/client.py:7]: Make it dynamic import

File: koreanbots/client.py [hunk 5]
```diff
@@ -1,251 +1,49 @@
...
-        예외
-        --------
-        .errors.HTTPException
-            알수없는 HTTP 에러가 발생했습니다, 주로 400에 발생합니다.
-        """
-        fp = await self.http.getVoteWidget(bot_id)
+class Koreanbots(KoreanbotsRequester):
+    def __init__(
+        self,
+        client: Optional[Type[discord.Client]],
```
Comment (kijk2869) [inline koreanbots/client.py:15]: client is not optional argument

File: koreanbots/client.py [hunk 6]
```diff
@@ -1,251 +1,49 @@
...
-        --------
-        .errors.HTTPException
-            알수없는 HTTP 에러가 발생했습니다, 주로 400에 발생합니다.
-        """
-        return await self.http.getVoteWidgetURL(bot_id)
-    
-    async def getServerWidget(self, bot_id: int, *args, filename='servers.png', **kwargs):
-        r"""주어진 봇ID의 서버 수 위젯의 discord.File을 가져옵니다.
+            with contextlib.suppress(HTTPException):
+                await self.guildcount(bot_id, total_guilds)
```
Comment (kijk2869) [inline koreanbots/client.py:32]: Should print or log the exception

File: koreanbots/client.py [hunk 7]
```diff
@@ -1,251 +1,49 @@
...
+    async def tasks_send_guildcount(self):
+        if not self.client:
+            raise RuntimeError("Client Not Found")
 
-        파라미터
-        -------------
-        bot_id: int
-            정보를 가져올 봇의 ID
+        await self.client.wait_until_ready()  # type: ignore
+        bot_id = (await self.client.application_info()).id  # type:ignore
```
Comment (kijk2869) [inline koreanbots/client.py:27]: Bot id will be loaded when ready

File: koreanbots/client.py [hunk 8]
```diff
@@ -1,251 +1,49 @@
...
-            반환받을 파일의 이름
+            await asyncio.sleep(1800)
 
-        예외
-        --------
-        .errors.HTTPException
-            알수없는 HTTP 에러가 발생했습니다, 주로 400에 발생합니다.
-        """
-        fp = await self.http.getVoteWidget(bot_id)
+    async def guildcount(self, bot_id: int, total_guilds: int):
```
Comment (kijk2869) [inline koreanbots/client.py:36]: Add return type

File: koreanbots/client.py [hunk 9]
```diff
@@ -1,251 +1,49 @@
...
+        return await self.post_update_bot_info(bot_id, total_guilds)
 
-        return discord.File(fp, *args, filename=filename, **kwargs)
-    
-    async def getServerWidgetURL(self, bot_id: int):
-        r"""주어진 봇ID의 서버 수 위젯 주소를 가져옵니다.
+    async def userinfo(self, user_id: int):
```
Comment (kijk2869) [inline koreanbots/client.py:39]: Add return type

File: koreanbots/client.py [hunk 10]
```diff
@@ -1,251 +1,49 @@
...
+        return await self.get_user_info(user_id)
 
-        파라미터
-        -------------
-        bot_id: int
-            정보를 가져올 봇의 ID
+    async def botinfo(self, bot_id: int):
```
Comment (kijk2869) [inline koreanbots/client.py:42]: Add return type

File: koreanbots/client.py [hunk 11]
```diff
@@ -1,251 +1,49 @@
...
-        --------
-        .errors.HTTPException
-            알수없는 HTTP 에러가 발생했습니다, 주로 400에 발생합니다.
-        """
-        return await self.http.getServerWidgetURL(bot_id)
\ No newline at end of file
+    @strict_literal("widget_type")
+    async def widget(
+        self, widget_type: Literal["votes", "servers", "status"], bot_id: int
+    ):
```
Comment (kijk2869) [inline koreanbots/client.py:48]: Add return type

File: koreanbots/decorator.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import functools
+from typing import Literal, get_args
+from koreanbots.http import KoreanbotsRequester
+from koreanbots.errors import AuthorizeError
```
Comment (kijk2869) [inline koreanbots/decorator.py:4]: Make it dynamic import

File: koreanbots/decorator.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+                literal_list = list(get_args(arg_annoration))
+                # Get index
+                arg_index = full_arg_spec.args.index(argument_name)
+                # Handle arguments
+                if arg_index < len(args) and args[arg_index] not in literal_list:
+                    raise ValueError(
+                        f"Arguments do not match. Expected: {literal_list}"
+                    )
+                # Handle keyword arguments
+                elif recive_arg := kwargs.get(argument_name):
```
Comment (kijk2869) [inline koreanbots/decorator.py:39]: This library supports **3.6+**

https://github.com/koreanbots/py-sdk/blob/fbb421f1ab9d2dc36a0dbcdd90b58c25a09427f7/setup.py#L40

File: koreanbots/errors.py [hunk 14]
```diff
@@ -22,49 +22,40 @@
...
-    """접근 권한이 없을 때 발생합니다."""
     pass
 
 
 class NotFound(HTTPException):
-    """해당 항목을 찾을 수 없을 때 발생합니다."""
     pass
+
+
+error_mapping = {400: BadRequest, 403: Forbidden, 404: NotFound}
```
Comment (kijk2869) [inline koreanbots/errors.py:61]:
```suggestion
ERROR_MAPPING = {400: BadRequest, 403: Forbidden, 404: NotFound}
```

File: koreanbots/http.py [hunk 15]
```diff
@@ -1,355 +1,65 @@
...
-        fp = BytesIO()
-        svg2png(bytestring=SVG, write_to=fp)
-        fp.seek(0)
-
-        return fp
+import aiohttp
+from typing import Any, Literal, Optional
+
+from koreanbots.decorator import required, strict_literal
+from koreanbots.errors import HTTPException, error_mapping
```
Comment (kijk2869) [inline koreanbots/http.py:5]: Make it dynamic import

File: koreanbots/http.py [hunk 16]
```diff
@@ -1,355 +1,65 @@
...
+BASE = "https://koreanbots.dev/api/"
+VERSION = "v2"
+
+KOREANBOTS_URL = BASE + VERSION
+
+
+class KoreanbotsRequester:
+    def __init__(
+        self,
+        api_key: Optional[str],
```
Comment (kijk2869) [inline koreanbots/http.py:16]: This argument is optional but there is no default argument

File: koreanbots/client.py [hunk 17]
```diff
@@ -1,251 +1,49 @@
...
-        --------
-        .errors.HTTPException
-            알수없는 HTTP 에러가 발생했습니다, 주로 400에 발생합니다.
-        """
-        fp = await self.http.getVoteWidget(bot_id)
+class Koreanbots(KoreanbotsRequester):
+    def __init__(
+        self,
+        client: Optional[Type[discord.Client]],
+        api_key: Optional[str],
```
Comment (kijk2869) [inline koreanbots/client.py:16]: This argument is optional but there is no default argument

File: koreanbots/client.py [hunk 18]
```diff
@@ -1,251 +1,49 @@
...
+        session: Optional[aiohttp.ClientSession],
```
Comment (kijk2869) [inline koreanbots/client.py:17]: This argument is optional but there is no default argument

File: koreanbots/client.py [hunk 19]
```diff
@@ -1,251 +1,49 @@
...
+    ) -> None:
```
Comment (kijk2869) [inline koreanbots/client.py:18]: Use typing.NoReturn

File: koreanbots/http.py [hunk 20]
```diff
@@ -1,355 +1,65 @@
...
+        session: Optional[aiohttp.ClientSession] = None,
+    ) -> None:
+        self.api_key = api_key
+        self.session = session
+
+    async def request(
+        self,
+        method: Literal["GET", "POST"],
+        return_method: Literal["json", "text", "read"],
```
Comment (kijk2869) [inline koreanbots/http.py:25]: I'm not suggesting this way.

File: koreanbots/http.py [hunk 21]
```diff
@@ -1,355 +1,65 @@
...
+    ):
+        if not self.session:
+            self.session = aiohttp.ClientSession()
+
+        async with self.session.request(
+            method, KOREANBOTS_URL + endpoint, **kwargs
+        ) as response:
+            if response.status != 200:
+                if error_mapping.get(response.status):
+                    raise error_mapping[response.status](
```
Comment (kijk2869) [inline koreanbots/http.py:37]:
```suggestion
                if ERROR_MAPPING.get(response.status):
                    raise ERROR_MAPPING[response.status](
```

File: koreanbots/http.py (hunk 15 above)
Comment (kijk2869) [inline koreanbots/http.py:5]:
```suggestion
from koreanbots.errors import HTTPException, ERROR_MAPPING
```

File: koreanbots/http.py [hunk 22]
```diff
@@ -1,355 +1,65 @@
...
+                        response.status, await response.json()
+                    )
+                else:
+                    raise HTTPException(response.status, await response.json())
+
+            return await getattr(response, return_method)()
+
+    @required
+    async def get_vote(self, user_id: int):
```
Comment (kijk2869) [inline koreanbots/http.py:46]: Add return type

File: koreanbots/http.py [hunk 23]
```diff
@@ -1,355 +1,65 @@
...
+        await self.request("GET", "json", f"/bots/voted/{user_id}")
+
+    async def get_bot_info(self, bot_id: int):
```
Comment (kijk2869) [inline koreanbots/http.py:49]: Add return type

File: koreanbots/http.py [hunk 24]
```diff
@@ -1,355 +1,65 @@
...
+class KoreanbotsRequester:
+    def __init__(
+        self,
+        api_key: Optional[str],
+        session: Optional[aiohttp.ClientSession] = None,
+    ) -> None:
+        self.api_key = api_key
+        self.session = session
+
+    async def request(
```
Comment (kijk2869) [inline koreanbots/http.py:22]: Apply strict literal

File: koreanbots/http.py [hunk 25]
```diff
@@ -1,355 +1,65 @@
...
+
+    @required
+    async def get_vote(self, user_id: int):
+        await self.request("GET", "json", f"/bots/voted/{user_id}")
+
+    async def get_bot_info(self, bot_id: int):
+        await self.request("GET", "json", f"/bots/{bot_id}")
+
+    @required
+    async def post_update_bot_info(self, bot_id: int, total_guilds: int):
```
Comment (kijk2869) [inline koreanbots/http.py:53]: Add return type

File: koreanbots/http.py [hunk 26]
```diff
@@ -1,355 +1,65 @@
...
+        await self.request(
+            "POST", "json", f"/bots/{bot_id}/stats", json={"servers": total_guilds}
+        )
+
+    @strict_literal("widget_type")
+    async def get_bot_widget(
+        self, widget_type: Literal["votes", "servers", "status"], bot_id: int
+    ):
```
Comment (kijk2869) [inline koreanbots/http.py:61]: Add return type

File: koreanbots/http.py [hunk 27]
```diff
@@ -1,355 +1,65 @@
...
+        await self.request("GET", "json", f"/widget/bots/{widget_type}/{bot_id}.svg")
```
Comment (kijk2869) [inline koreanbots/http.py:62]: Add a way to use png

File: koreanbots/client.py [hunk 28]
```diff
@@ -1,251 +1,49 @@
...
+        self.client = client
+        super().__init__(api_key, session=session)
```
Comment (kijk2869) [inline koreanbots/client.py:20]: Change the order

**Review by kijk2869 — CHANGES_REQUESTED** [review]

**Comment by SaidBySolo** [comment]
TODO:
Refactor rate-limit handler

.pop -> .get

File: koreanbots/model.py [hunk 29]
```diff
@@ -1,119 +1,143 @@
...
+    @property
+    def banner(self) -> Optional[str]:
+        return self.data.get("banner", None)
+
+    @property
+    def status(self) -> Optional[Status]:
+        return self.data.get("status", None)
+
+    @property
+    def state(self) -> Optional[State]:
```
Comment (wonderlandpark) [inline koreanbots/model.py:113]: Not nullable
```suggestion
    def state(self) -> State:
```

**Review by wonderlandpark — CHANGES_REQUESTED** [review]

**Comment by wonderlandpark** [comment]
Needs to support vote check api
(Added to docs)
@SaidBySolo sa

File: koreanbots/typing.py [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+WidgetStyle = Literal["classic", "flat"]
+
+Category = Literal[
+    "관리",
+    "뮤직",
+    "전적",
+    "게임",
+    "도박",
+    "로깅",
+    "슬래시 명령어",
```
Comment (wonderlandpark) [inline koreanbots/typing.py:22]:
```suggestion
    "빗금 명령어",
```

**Review by wonderlandpark — CHANGES_REQUESTED** [review]

**Comment by wonderlandpark** [comment]
Will only support `svg` format for the widget.
Needs action for it. Should convert it to png, jpg, webp etc.

**Review by zero734kr — COMMENTED** [review]
I will review this PR later; just commenting to show this at my main page

**Comment by wonderlandpark** [comment]
> I will review this PR later; just commenting to show this on my main page

Requsted review

File: koreanbots/decorator.py [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+                literal_list = list(get_args(arg_annoration))
+                # Get index
+                arg_index = full_arg_spec.args.index(argument_name)
+                # Handle arguments
+                if arg_index < len(args) and args[arg_index] not in literal_list:
+                    raise ValueError(
+                        f"Arguments do not match. Expected: {literal_list}"
+                    )
+                # Handle keyword arguments
+                elif kwargs.get(argument_name):
```
Comment (zero734kr) [inline koreanbots/decorator.py:43]:
```suggestion
                elif kwargs.get(argument_name) and kwargs[argument_name] not in literal_list:
```

File: koreanbots/errors.py [hunk 32]
```diff
@@ -1,70 +1,46 @@
...
-    """koreanbots의 기본 예외 클래스입니다.
-    즉, 이 라이브러리의 모든 예외는 이 예외로 잡을 수 있습니다.
-    """
+from typing import Any, Dict, TYPE_CHECKING, Union
+
+if TYPE_CHECKING:
+    from .typing import ErrorMapping
+
+
+class KoreanbotsExeption(Exception):
```
Comment (zero734kr) [inline koreanbots/errors.py:7]: Typo here
```suggestion
class KoreanbotsException(Exception):
```

File: koreanbots/errors.py [hunk 33]
```diff
@@ -1,70 +1,46 @@
...
-        self.status = response.status
+
+class HTTPException(KoreanbotsExeption):
+    def __init__(self, code: Any, message: Union[Any, Dict[str, Any]]):
+        self.status = code
         if isinstance(message, dict):
-            self.status = message.get('code', self.status)
-            self.error = message.get('message', 'DBKRException')
+            self.status = message.get("code", self.status)
+            self.error = message.get("message", "KoreanbotsExeption")
```
Comment (zero734kr) [inline koreanbots/errors.py:20]: Typo

**Review by zero734kr — CHANGES_REQUESTED** [review]

**Review by wonderlandpark — CHANGES_REQUESTED** [review]
Could you update this URL?
https://github.com/koreanbots/py-sdk/blob/5e148cdb8f522cf20dd4c83f828a34f1f14f3e57/setup.py#L28

File: setup.py [hunk 34]
```diff
@@ -22,13 +20,6 @@
 
 setup(
     name='koreanbots',
-    author='kijk2869',
-    url='https://github.com/kijk2869/koreanbots',
-    project_urls={
-        "Homepage": "https://koreanbots.cf/",
-        "Source":"https://github.com/koreanbots/py-sdk",
-        "Tracker":"https://github.com/koreanbots/py-sdk/issues"
-    },
```
Comment (kijk2869) [inline setup.py:31]: Restore it with refreshed information

File: setup.py [hunk 35]
```diff
@@ -3,9 +3,7 @@
 from setuptools import setup
 import re, os
 
-version = ''
-with open('koreanbots/__init__.py', encoding='UTF8') as f:
-    version = re.search(r'^__version__\s*=\s*[\'"]([^\'"]*)[\'"]', f.read(), re.MULTILINE).group(1)
+version = '2.0.0'
```
Comment (kijk2869) [inline setup.py:6]: Save version on __init__.py for version information to be accessible on inline code.

**Review by kijk2869 — CHANGES_REQUESTED** [review]

File: koreanbots/abc.py [hunk 36]
```diff
@@ -0,0 +1,20 @@
+import abc
+from typing import Any, Dict
+
+
+class KoreanbotsABC(metaclass=abc.ABCMeta):
+    @abc.abstractmethod
+    def __init__(self, **response_data: Any) -> None:
```
[the remark this answers is not in the document]
Comment (SaidBySolo) [reply koreanbots/abc.py:7]: Can't use kwarg in TypedDict

File: koreanbots/model.py [hunk 37]
```diff
@@ -1,119 +1,167 @@
...
-)
\ No newline at end of file
+from typing import Any, Dict, List, Optional, Union
+
+from .abc import KoreanbotsABC
+from .typing import Category, State, Status
+
+
+class BaseKoreanbots(KoreanbotsABC):
+    def __init__(self, **response_data: Any) -> None:
```
[the remark this answers is not in the document]
Comment (SaidBySolo) [reply koreanbots/model.py:8]: Can't use TypedDict in kwarg

**Review by wonderlandpark — APPROVED** [review]

**Pull request merged**